crates/ironrdp-rdpdr-native/src/nix/simple.rs

```diff
@@ -0,0 +1,790 @@
+                            },
+                        ))])
+                    } else {
+                        warn!("unsupported file class for query directory");
+                        Ok(vec![SvcMessage::from(RdpdrPdu::ClientDriveQueryDirectoryResponse(
+                            ClientDriveQueryDirectoryResponse {
+                                device_io_reply: DeviceIoResponse::new(device_io_request, NtStatus::NOT_SUPPORTED),
+                                buffer: None,
+                            },
+                        ))])
+                    }
+                }
+                Err(error) => {
+                    warn!(%error, "Get metadata error");
+                    Ok(vec![SvcMessage::from(RdpdrPdu::ClientDriveQueryDirectoryResponse(
+                        ClientDriveQueryDirectoryResponse {
+                            device_io_reply: DeviceIoResponse::new(device_io_request, not_found_status),
+                            buffer: None,
+                        },
+                    ))])
+                }
+            }
+        }
+    }
+}
+pub(crate) fn query_directory(
+    backend: &mut SimpleRdpdrBackend,
+    req_inner: ServerDriveQueryDirectoryRequest,
+) -> PduResult<Vec<SvcMessage>> {
+    match backend.file_path_map.get(&req_inner.device_io_request.file_id) {
+        Some(parent_pos_for_next) => {
+            let mut find_file_name = None;
+            if req_inner.initial_query > 0 {
+                if req_inner.path.ends_with('*') {
+                    let mut parent = backend.file_base.clone();
+                    let query_path = req_inner.path.replace('\\', "/");
+                    let len = query_path.len();
+                    // path ends with *, so its len > 0
+                    #[allow(clippy::arithmetic_side_effects)]
+                    parent.push_str(&query_path[0..len - 1]);
+                    let dirp = if let Ok(real_parent_c) = CString::new(parent.clone()) {
+                        // SAFETY: it is safe to call this on *nix platform
+                        unsafe { opendir(real_parent_c.as_ptr()) }
+                    } else {
+                        std::ptr::null::<DIR>().cast_mut()
+                    };
+                    if !dirp.is_null() {
+                        loop {
+                            // SAFETY: it is safe to call readdir on *nix platform
+                            let first = unsafe { readdir(dirp) };
+                            if first.is_null() {
+                                break;
+                            }
+                            // SAFETY: it is safe to call this for rdp
+                            let dir_name = unsafe { &(*first).d_name };
+                            // SAFETY: it is safe to call this for rdp
+                            let dir_name_in_rs =
+                                unsafe { CStr::from_ptr(dir_name as *const i8).to_string_lossy().into_owned() };
```

Comment:
question: What do you mean by "for rdp"? This seems unrelated to the RDP protocol.

suggestion: In this case, it’s a FFI structure and we can only assume that the library is well-behaved.
```suggestion
                            // SAFETY: When a non-null value is returned, the pointed struct should be properly initialized by the library.
                            let dir_name = unsafe { &(*first).d_name };

                            // SAFETY: DIR::name field contains the null terminated filename.
                            // etc etc
                            let dir_name_in_rs =
                                unsafe { CStr::from_ptr(dir_name as *const i8).to_string_lossy().into_owned() };
```

Also check the safety section of `CStr::from_ptr`: https://doc.rust-lang.org/std/ffi/struct.CStr.html#safety

suggestion: You can also likely encapsulate this in `BoxDir` directly.